src/main/java/software/amazon/smithy/lsp/ext/SmithyProject.java

```diff
@@ -159,37 +161,162 @@ public File getRoot() {
+                    if (!traits.isEmpty()) {
+                        traits.sort(new SourceLocationSorter());
+                        currentMemberEndMarker = traits.get(0).getSourceLocation().getLine();
+                    }
+                    memberEndMarker = currentMemberEndMarker - 1;
+                } else {
+                    // When handling non-member shapes, advance the end marker for traits and comments above the current
+                    // shape.
+                    endMarker = startPosition.getLine();
+                    List<Trait> traits = new ArrayList<>(shape.getAllTraits().values());
+                    // If the shape has traits, advance the end marker again.
+                    if (!traits.isEmpty()) {
+                        // TODO: Replace with Comparator when this class is removed.
+                        traits.sort(new SourceLocationSorter());
+                        endMarker = traits.get(0).getSourceLocation().getLine() - 1;
+                    }
+                    // Move the end marker when encountering line comments or empty lines.
+                    if (lines.size() > endMarker) {
+                        while (lines.get(endMarker - 1).trim().startsWith("//")
+                                || lines.get(endMarker - 1).trim().equals("")) {
+                            endMarker = endMarker - 1;
+                        }
+                    }
+                }
+                Location location =  new Location(getUri(modelFile), new Range(startPosition, endPosition));
+                locations.put(shape.getId(), location);
+            }
+        }
         return locations;
     }
 
+    /**
+     * Returns the shapeId of the shape that corresponds to the file uri and position within the model.
+     *
+     * @param uri String uri of model file
+     * @param position Cursor position within model file
+     * @return ShapeId of corresponding shape defined at location.
+     */
+    public Optional<ShapeId> getShapeIdFromLocation(String uri, Position position) {
+        Comparator<Map.Entry<ShapeId, Location>> rangeSize = Comparator.comparing(entry ->
+                entry.getValue().getRange().getEnd().getLine() - entry.getValue().getRange().getStart().getLine());
+
+        List<Map.Entry<ShapeId, Location>> matchingShapes = locations.entrySet().stream()
+                .filter(entry -> entry.getValue().getUri().endsWith(Paths.get(uri).toString()))
+                .filter(filterByLocation(position))
+                // Since the position is in each of the overlapping shapes, return the smallest range.
+                .sorted(rangeSize)
+                .collect(Collectors.toList());
+        if (matchingShapes.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(matchingShapes.get(0).getKey());
+    }
+
+    private Predicate<Map.Entry<ShapeId, Location>> filterByLocation(Position position) {
```

Comment:
This seems like it should be a `Predicate<Location>` instead, since it doesn't use the key at all.